### PR TITLE
test(e2e): add a Gatsby Preview e2e test

### PR DESCRIPTION
**Pull request by DSchau**
## Description

This builds upon the development Cypress e2e test, and adds a few features, specifically:

1. A (mock/fake) gatsby-source-* plugin to mimic a real source plugin (e.g. Contentful, Sanity, etc.) that offers a "sync" API to also delete nodes when reset
1. A suite of e2e tests around Gatsby Preview functionality (e.g. content appearing when refreshed, content deleting when reset, etc.)

I also added `cypress-testing-library` to get `queryByText` support, which we should _probably_ prefer over data-testid going forward. Happy to add it to others, but don't think it should block this PR.

~There is one failing test I'm still ironing out, so I'll fix that and remove the draft/WIP.~

### CODE REVIEW

File: e2e-tests/development-runtime/cypress/integration/gatsby-preview/updating.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    cy.get(`li:eq(0) a`)
+      .click()
+      .waitForRouteChange()
+
+    update()
+
+    cy.queryByText(`1`).should(`exist`)
+  })
+
+  it.skip(`updates and hot-reloads new content`, () => {
```
Comment (DSchau) [inline e2e-tests/development-runtime/cypress/integration/gatsby-preview/updating.js]: This intermittently fails, not sure why exactly.

File: e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    const node = getNode(
+      {
+        updates: 0,
+        uuid,
+        title: `Hello World (${uuid})`,
+        message: `
+This is fake data to test Gatsby Preview.
+What are you doing here, anyways? Be gone!
+
+(also pretend this is randomized, it is not)
```
Comment (sidharthachatterjee) [inline e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js]: 😄

File: e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+...
+
+OK cool, I appreciate you. Thanks for checking deep into the underlying stuff, you pro.
```
Comment (sidharthachatterjee) [inline e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js]: ❤️

File: e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+module.exports = {
+  dbFilePath: path.join(__dirname, `db.json`),
+  nodeType: NODE_TYPE,
+  nodes: [],
+  async sync(helpers) {
+    // empty array
+    const nodes = await readFile(this.dbFilePath)
+
+    const uuid = max(nodes) + 1
```
Comment (sidharthachatterjee) [inline e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js]: Nit: Could use `uuid`

File: e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    .promisify(fs.readFile)(file, `utf8`)
+    .then(contents => JSON.parse(contents))
+    .catch(() => [])
+
+const writeFile = (file, contents) =>
+  util
+    .promisify(fs.writeFile)(file, JSON.stringify(contents), `utf8`)
+    .catch(() => null)
+
+const difference = (updated, existing) => {
```
Comment (sidharthachatterjee) [inline e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js]: Nit: Use `_.difference`?

File: e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+const readFile = file =>
+  util
+    .promisify(fs.readFile)(file, `utf8`)
+    .then(contents => JSON.parse(contents))
+    .catch(() => [])
+
+const writeFile = (file, contents) =>
+  util
+    .promisify(fs.writeFile)(file, JSON.stringify(contents), `utf8`)
+    .catch(() => null)
```
Comment (sidharthachatterjee) [inline e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js]: Nit: Log and then swallow error 😛

File: e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js (hunk 5 above)
Comment (DSchau) [reply e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js]: We use lodash too often, as is ;) 🔥

File: e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js (hunk 4 above)
Comment (DSchau) [reply e2e-tests/development-runtime/plugins/gatsby-source-fake-data/api.js]: Eh - it functions as a uuid, don't see the need for an extra dep!

**Pull request merged**